activerecord/lib/active_record/autosave_association.rb

```diff
@@ -273,7 +282,23 @@ def changed_for_autosave?
       new_record? || has_changes_to_save? || marked_for_destruction? || nested_records_changed_for_autosave?
     end
 
+    # Returns whether or not this record is already being saved outside of the
+    # current autosave callback
+    def already_saving?
+      @saving
+    end
+
     private
+      # Track if this record is already being saved. If it is being saved we
+      # can skip saving it in the autosave callbacks
+      def with_saving_state_tracking
+        @saving = true
+        status = yield
+      ensure
+        @saving = false
+        status
+      end
```

Comment:
Don’t need the `status` variable here. Ensure blocks don’t affect return value.

```
irb(main):001:1* def foo
irb(main):002:1*   :foo
irb(main):003:1* ensure
irb(main):004:1*   :bar
irb(main):005:1* end
irb(main):006:0> foo
=> :foo
```

I think there’s a risk that this method may be called while already saving, too (if an app callback on a child record explicitly saves its parent record). Should we track previous saving state on the stack to be safe?

```suggestion
      # Track if this record is already being saved. If it is being saved we
      # can skip saving it in the autosave callbacks
      def with_saving_state_tracking
        previously_saving = @saving
        @saving = true
  
        yield
      ensure
        @saving = previously_saving
      end
```

Reply:
Ah, didn't think of that. That seems like a good idea.
I'll change it and see if I can add a test for this behaviour.